### PR TITLE
Add AWS DynamoDB Streams Scaler

### LINKED ISSUE #3124: Add AWS DyanamoDB Streams Scaler

### PR DESCRIPTION
**Pull request by yokawasa**
Signed-off-by: Yoichi Kawasaki <yokawasa@gmail.com>

_Provide a description of what has been changed_

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Tests have been added
- [x] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)

Fixes #3124

Relates to #

docs - https://github.com/kedacore/keda-docs/pull/771

### CODE REVIEW

File: tests/scalers/aws-dynamodb-stream.test.ts
Comment (JorTurFer) [inline tests/scalers/aws-dynamodb-stream.test.ts:73]: shouldn't this fail if shard count is 0?

**Review by JorTurFer — COMMENTED** [review]
Looking good in general!
Thanks for this new scaler! ❤️ 
I have left some comments inline and apart from that, I'm a bit worried about the e2e test. 
Right now, they are quite "random" because the e2e test only checks that from 1 it goes to `dynamoDBStreamShardNum` (which could be 1). I'd change a bit the test to ensure the behaviour. 
I mean, e2e test could start from 0, that will validate than at least we are getting the value and scale from 0 to 1. Other nice check could be scaling out/in, I have checked that shard count is nor modifiable easily, but maybe we can modify the `shardCount` in the SO, something like:

1.  0 instances, deploy the SO with `shardCount: dynamoDBStreamShardNum` and check that it goes to 1
2. 1 instance, deploy the SO with `shardCount: dynamoDBStreamShardNum * 2` and check that it goes to 2
3. 2 instances, deploy the SO with `shardCount: dynamoDBStreamShardNum` and check that it goes to 1

There is an example of this approach with [memory scaler e2e tests](https://github.com/kedacore/keda/blob/main/tests/scalers/memory.test.ts#L119-L143) 
WDYT?

The last thing, [static check](https://github.com/kedacore/keda/runs/6710532495?check_suite_focus=true) is failing, could you take a look?

File: tests/scalers/aws-dynamodb-stream.test.ts
Comment (yokawasa) [reply tests/scalers/aws-dynamodb-stream.test.ts:73]: As you say, this should fail if the count is 0.  
So I added the check like this https://github.com/kedacore/keda/pull/3121/commits/593eebbc4032c5de9529919c11e32e270a626b0e

File: pkg/scalers/aws_dynamodb_stream_scaler.go
Comment (yokawasa) [reply pkg/scalers/aws_dynamodb_stream_scaler.go]: I modified the code to use `DescribeTableWithContext` https://github.com/kedacore/keda/pull/3121/commits/98d2fa45938929879db841346b91dec221f79dc2

File: pkg/scalers/aws_dynamodb_stream_scaler.go
Comment (yokawasa) [reply pkg/scalers/aws_dynamodb_stream_scaler.go]: I modified the code to use `DescribeStreamWithContext ` https://github.com/kedacore/keda/pull/3121/commits/98d2fa45938929879db841346b91dec221f79dc2

**Comment by yokawasa** [comment]
@JorTurFer 

> The last thing, [static check](https://github.com/kedacore/keda/runs/6710532495?check_suite_focus=true) is failing, could you take a look?

I fixed! https://github.com/kedacore/keda/pull/3121/commits/d0100b0502ad8a8e15a3442b643921c3a2240175. I no longer see static check error - [static cehck](https://github.com/kedacore/keda/runs/6724656040?check_suite_focus=true)

> Right now, they are quite "random" because the e2e test only checks that from 1 it goes to dynamoDBStreamShardNum > (which could be 1). I'd change a bit the test to ensure the behaviour.
> I mean, e2e test could start from 0, that will validate than at least we are getting the value and scale from 0 to 1. Other nice check could be scaling out/in, I have checked that shard count is nor modifiable easily, but maybe we can modify the shardCount in the SO, something like:
> 
> 0 instances, deploy the SO with shardCount: dynamoDBStreamShardNum and check that it goes to 1
> 1 instance, deploy the SO with shardCount: dynamoDBStreamShardNum * 2 and check that it goes to 2
> 2 instances, deploy the SO with shardCount: dynamoDBStreamShardNum and check that it goes to 1
> There is an example of this approach with [memory scaler e2e tests] (https://github.com/kedacore/keda/blob/main/tests/scalers/memory.test.ts#L119-L143)
> WDYT?

Thank you for the suggestions!
`shardCount` is the target value that a DynamoDB streams consumer can handle, thus deploying the SO with shardCount: dynamoDBStreamShardNum * 2 doesn't lead the replicas count go to 2. A problem is like you say the DynamoDB stream shard count is not modifiable easily. 
I come up with the following tests. 

- test1 (scale out check):  0 instances, deploy the SO with shardCount: dynamoDBStreamShardNum and check that it goes to 1
- test2 (scale in check):  1 instance, manually scale up the instance # (replicas #) to dynamoDBStreamShardNum * 2, and check that it goes to 1 ( the SO stay with shardCount: dynamoDBStreamShardNum)

What do you think?

**Comment by JorTurFer** [comment]
you are right, `dynamoDBStreamShardNum * 2` will not work :(
Is there any option of generating load to make shardCount increases? I mean, maybe with one job running in the cluster and sending data to the DynamoDB Stream, with any consumer maybe we can increase the shardCount. (or maybe not).
In case of not, do you know the amount of shard that it has usually? I mean, if it has minimum 2 shards, we could set target to 1 and to 2...
If nothing of the previous options works, scaling from zero to 1 it should be enough taking into account that nothing else is possible

**Comment by yokawasa** [comment]
@JorTurFer 
Thank you for the comment!!

> Is there any option of generating load to make shardCount increases? 

We can generate load to make shardCount increases but it's hard to know time frame for it and when it will increase. 

> In case of not, do you know the amount of shard that it has usually? I mean, if it has minimum 2 shards, we could set target to 1 and to 2...

I believe it's not officially documented but as far as I know, a stream initally has at least 2 (usually 4) shards if the dynamodb table to which the stream belongs is created with BillingMode: `PAY_PER_REQUEST` (vs. `PROVISIONED`).
So as your suggestion, we can set target to 1 and to 2...

Create a test dynamodb table with BillingMode: `PAY_PER_REQUEST`
0 instances, deploy the SO with shardCount: dynamoDBStreamShardNum and check that it goes to 1
1 instance, deploy the SO with shardCount: 1  and check that it goes to the same number of `dynamoDBStreamShardNum`
N (N>1) instances, deploy the SO with shardCount: dynamoDBStreamShardNum and check that it goes back to 1

I'll modify like this if it sounds good for you

**Comment by JorTurFer** [comment]
sounds good! 😄
Thanks a lot for the extra effort you are doing! 🙇

**Comment by yokawasa** [comment]
@JorTurFer  

> Create a test dynamodb table with BillingMode: PAY_PER_REQUEST and stream enabled
> 0 instances, deploy the SO with shardCount: dynamoDBStreamShardNum and check that it goes to 1
> 1 instance, deploy the SO with shardCount: 1 and check that it goes to 2 (=maxReplicaCount 2)
> 2 instances, deploy the SO with shardCount: dynamoDBStreamShardNum and check that it goes back to 1

thank you for the check. I modify the e2e test https://github.com/kedacore/keda/pull/3121/commits/85e553cb9953a0e477a447458d730f0f3b6a9c81

**Comment by JorTurFer** [comment]
Looks fantastic. 
Remember please to add an entry (alphabetically sorted) in the changelog (The issue you have to refer is this `https://github.com/kedacore/keda/issues/3124`) under the section unreleased/new.
Something like:
`- **General:** Introduce new AWS DynamoDB Streams Scaler ([#3124](https://github.com/kedacore/keda/issues/3124))`

File: pkg/scalers/aws_dynamodb_streams_scaler.go
Comment (zroubalik) [inline pkg/scalers/aws_dynamodb_streams_scaler.go:136]: Just double checking, these clients doesn't have to be closed?

File: tests/scalers/aws-dynamodb-streams.test.ts
Comment (zroubalik) [inline tests/scalers/aws-dynamodb-streams.test.ts:1]: Sorry for being pain, right now, we are in the process in migrating TS based e2e tests to Go. Do you think you can rewrite this test to use Go?

You can follow this https://github.com/kedacore/keda/pull/3079, and once it is merged, you can use it as a base for this test. 

Thanks!

File: tests/scalers/aws-dynamodb-streams.test.ts
Comment (zroubalik) [inline tests/scalers/aws-dynamodb-streams.test.ts:12]: This namespace is too generic, please add some prefix so it is unique for this test.

e2e tests are running in the parallel, so this might cause problems.

File: tests/scalers/aws-dynamodb-streams.test.ts
Comment (zroubalik) [inline tests/scalers/aws-dynamodb-streams.test.ts:9]:
```suggestion
const awsRegion = 'eu-west-2'
```

please use this region, so we are in sync with other AWS tests we have.

File: pkg/scalers/aws_dynamodb_streams_scaler.go
Comment (yokawasa) [reply pkg/scalers/aws_dynamodb_streams_scaler.go:136]: No, it doesn't have to be closed

File: tests/scalers/aws-dynamodb-streams.test.ts
Comment (yokawasa) [reply tests/scalers/aws-dynamodb-streams.test.ts:1]: @zroubalik Yes. I'm happy to rewrite the test in Golang!

File: pkg/scalers/aws_dynamodb_streams_scaler.go
Comment (zroubalik) [reply pkg/scalers/aws_dynamodb_streams_scaler.go:136]: Ack :)

File: tests/scalers/aws-dynamodb-streams.test.ts
Comment (zroubalik) [reply tests/scalers/aws-dynamodb-streams.test.ts:1]: Awesome, thanks. So let's wait until the e2e tests migration PR is merged (it should be hopefully by the end of this week).

Comment (zroubalik) [reply tests/scalers/aws-dynamodb-streams.test.ts:1]: @yokawasa FYI  he initial work on Golang based e2e tests has been done, you can try to follow the same pattern now.

Comment (yokawasa) [reply tests/scalers/aws-dynamodb-streams.test.ts:1]: @zroubalik Thank you for the instruction! I've added Go based e2e tests for dynamodb streams scaler. Please reivew it https://github.com/kedacore/keda/pull/3121/commits/5ca8d02a97b8890580ea385c6b8d7011022f5707

Comment (JorTurFer) [reply tests/scalers/aws-dynamodb-streams.test.ts:1]: Could you remove the tests in ts? the idea is having only in Golang so doesn't make sense adding also in ts

**Review by JorTurFer — APPROVED** [review]

File: tests/scalers_go/aws_dynamodb_streams/aws_dynamodb_streams_test.go
Comment (JorTurFer) [inline tests/scalers_go/aws_dynamodb_streams/aws_dynamodb_streams_test.go:28]:
```suggestion
	. ""github.com/kedacore/keda/v2/tests/helper"
```

**Comment by yokawasa** [comment]
@JorTurFer @zroubalik sorry for my slow update. I've fixed all conflicts and link errors. Pleae review again 🙇

**Review by zroubalik — APPROVED** [review]
LGTM, thanks for the contribution!

**Pull request merged**
